### ISSUE TITLE
abap <-> json name mapping not working when converting json to abap table

### ISSUE THREAD

**Issue by xrevilla** [se38/zJSON#34]
Hi, first of all, thanks a lot for this great and useful project. 

I think I found a bug when trying to convert a complex json to an abap table. In my case it fails and does not return the result (but no error is raised).

I have this json:
[{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T10:47:34.983","status":"alta","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10"],"rights":[]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":51},{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T10:50:36.677","status":"bajaa","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10"],"rights":[]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":52},{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T11:10:41.56","status":"baja","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10"],"rights":[]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":53}]

And this complex abap type:

    TYPES: BEGIN OF ts_rights,
             info TYPE string,
           END OF ts_rights,
           tt_rights TYPE STANDARD TABLE OF ts_rights
                          WITH NON-UNIQUE DEFAULT KEY.
    TYPES: BEGIN OF ts_purchase_info,
             email    TYPE string,
             password TYPE string,
             name     TYPE string,
             surname  TYPE string,
           END OF ts_purchase_info.
    TYPES: BEGIN OF ts_purchase_products,
             crm_id        TYPE string,
             description   TYPE string,
             price         TYPE string,
             purchase_date TYPE string,
             end_date      TYPE string,
             app_id        TYPE string,
             app_skus      TYPE tt_app_skus,
             rights        TYPE tt_rights,
           END OF ts_purchase_products,
           tt_purchase_products TYPE STANDARD TABLE OF ts_purchase_products
                                     WITH NON-UNIQUE DEFAULT KEY.
    TYPES: BEGIN OF ts_purchase,
             id                TYPE string,
             crm_id            TYPE string,
             user_id           TYPE string,
             purchase_date     TYPE string,
             update_date       TYPE string,
             status            TYPE string,
             price             TYPE string,
             purchase_products TYPE tt_purchase_products,
             purchase_info     TYPE ts_purchase_info,
           END OF ts_purchase,
           tt_purchase TYPE STANDARD TABLE OF ts_purchase
                            WITH NON-UNIQUE DEFAULT KEY.

When I transform the json to a variable type tt_purchase it returns nothing. Code snippet:

      TRY.
          DATA(lo_json) = zcl_json_document=>create_with_json( json = lv_respuesta
                                                               name_mappings = VALUE #( ( abap_name = 'CRM_ID'            json_name = 'crmId' )
                                                                                        ( abap_name = 'USER_ID'           json_name = 'userId' )
                                                                                        ( abap_name = 'PURCHASE_DATE'     json_name = 'purchaseDate' )
                                                                                        ( abap_name = 'END_DATE'          json_name = 'endDate' )
                                                                                        ( abap_name = 'UPDATE_DATE'       json_name = 'updateDate' )
                                                                                        ( abap_name = 'PURCHASE_PRODUCTS' json_name = 'purchaseProducts' )
                                                                                        ( abap_name = 'PURCHASE_INFO'     json_name = 'purchaseInfo' )
                                                                                        ( abap_name = 'APP_ID'            json_name = 'appId' )
                                                                                        ( abap_name = 'APP_SKUS'          json_name = 'appSkus' ) ) ).
          lo_json->get_data(
            IMPORTING
              data = et_purchase
          ).
        CATCH zcx_json_document INTO DATA(lx_json_document).
          cl_message_helper=>set_msg_vars_for_if_t100_msg( lx_json_document ).
          eo_log->add_message( VALUE #( id = sy-msgid
                                        type = zcx_generic=>gc_message_type-error
                                        number = sy-msgno
                                        message_v1 = sy-msgv1
                                        message_v2 = sy-msgv2
                                        message_v3 = sy-msgv3
                                        message_v4 = sy-msgv4 ) ).
      ENDTRY.

Where lv_respuesta contains the json and et_purchase is the abap table.

I made the following corrections in the private method GET_TABLE of the class ZCL_JSON_DOCUMENT and now its working:
![image](https://user-images.githubusercontent.com/59660931/156189767-a2923375-5efe-4609-99cc-f273431d141b.png)
 When obtaining the values of the structure for the resulting table, it was not applying the mapping from abap to json name (in the method GET_STRU is aplying it).

I post this here in case I'm calling the class in a wrong way and this is really not a bug (but it seems so).

**Comment by se38** [comment]
@xrevilla can you please give me the definition of "tt_app_skus" to test with your example?

**Comment by xrevilla** [comment]
> @xrevilla can you please give me the definition of "tt_app_skus" to test with your example?

Yes, sorry, it's the following:
    TYPES: tt_app_skus TYPE STANDARD TABLE OF string
                            WITH NON-UNIQUE DEFAULT KEY.

Greetings!

**Comment by se38** [comment]
I've created a new branch named "[issue-34-name-mappings](https://github.com/se38/zJSON/tree/issue-34-name-mappings)" with your suggestions (with a little difference because of downward compatibility).

Your example works with your test data.
Can you please check with additional data?

**Comment by xrevilla** [comment]
Hello @se38 ,

I've tested with your changes and it works without any problems. I've tested also with a modified version of the json in my first example:
[{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T10:47:34.983","status":"alta","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10"],"rights":[]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":51},{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T10:50:36.677","status":"bajaa","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10", "es.test.test.11", "es.test.test.12", "es.test.test.13"],"rights":[]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":52},{"crmId":"CRM2022-0011","userId":"718e9fa4-599a-47b6-b66a-2673e30941d4","purchaseDate":"2022-02-25T00:00:00","updateDate":"2022-02-25T11:10:41.56","status":"baja","price":1,"purchaseProducts":[{"crmId":"JUE2","description":"El Test Digital","price":1,"purchaseDate":"2022-02-25T10:27:19.222Z","endDate":"2022-02-25T10:27:19.222Z","appId":"elTest","appSkus":["es.test.test.10"],"rights":[{"info":"Info Test"}]}],"purchaseInfo":{"email":"test@test.com","password":null,"name":"Test","surname":"Test"},"id":53}]

This also works ok (with rights and more entries in appSkus).

To me it's working well and can be closed if you agree.

Thanks!
